articles/2020-lkingsford_writing-your-first-unit-test-with-pytest.md

```diff
@@ -0,0 +1,386 @@
+FAILED tests/test_user.py::test_get_display_name - AssertionError: Display name should have returned preferred name
+================================================== 1 failed in 0.08s ===================================================
+```
+
+Unit tests will usually fail because of an `assert` that failed, or an exception
+that was thrown during the test. As you can see here, you'll see that a test
+failed, and a stack trace as to how.
+
+If your project uses Python's `logging` library, any log messages that are a
+higher level than `WARNING` will be shown with any failing tests.
+
+## Parametrizing Tests
+
+Before we start, we're going to need to fix `User.get_display_name` (since we
+intentionally broke the model in the last test). Change `User.get_display_name`
+in `models/user.py` back to the following:
+
+```python
+    def get_display_name(self):
+        """Return the name to display for this user.
+
+        The users display name is their preferred_name if it is set,
+        if not, their real_name if it is set, alternatively, if not,
+        their username"""
+        if self.preferred_name:
+            return self.preferred_name
+        if self.real_name:
+            return self.real_name
+        return self.username
+```
+
+Our first test only tested one specific part of the specification - namely that
+`preferred_name` would be returned if `User` has a `real_name`, `preferred_name`
+and `username`. This obviously would not pick up some bugs. For instance, if
+the function wrongly *always* returned `preferred_name` (like we forced it to
+always return `username` when we intentionally broke it) then that unit test
+would pass.
+
+The obvious solution is that we need to duplicate the test a bunch of times
+with all of the potential combinations. So - with over-application of our
+'Copy' and 'Paste' functions, we'll end up with four different copies of our
+test function, that could test the entire spec. That's not too ludicrous,
+but there is a better way.
+
+After all - if you want to test your string manipulation function with the
+entire [Big List of Naughty Strings](https://github.com/minimaxir/big-list-of-naughty-strings)
+you probably don't want to make a few hundred copies of the same function.
+
+Lots of testing frameworks - including Pytest - support a feature called
+'parametrization'. This means that can get data from an iterator and run a
+test with that data
+
+> Yes, you can use generators - as your can use any iterators. It's outside the
+> scope of this article to discuss exactly how. The official documentation for
+> Pytest has a [pretty in-depth look at how you can use parametrization](http://doc.pytest.org/en/latest/example/parametrize.html)
+> that is worth a read.
+
+So - here is our new version of `tests/test_user.py` that uses parametrization
+to test all combinations of input.
+
+```python
+import pytest
+import models.user
+
+
+TEST_DISPLAY_NAME_DATA = [
+    ['username', None,None, 'username'],
```

Comment:
Should you have tuples for these values?
```suggestion
    ('username', None,None, 'username'),
```

Reply:
It can be either. Maybe I'll make it tuples for more clarity in writing.